docs/architecture/joins.txt

```diff
@@ -163,17 +160,123 @@ earlier:
 Consequently, the number of rows is significantly reduced allowing the
 operation to complete much faster.
 
+.. _joins_hash_join:
+
+Hash Join
+---------
+
+The Hash Join algorithm is used to execute certains types of joins in a more
+perfomant way than :ref:`Nested Loop <joins_nested_loop>`.
+
+Basic Algorithm
+...............
+
+The operation takes place in one node (the handler node to which the client is
+connected). The rows of the left relation of the join are read and a hashing
+algorithm is applied on the fields of the relation which participate in the
+join condition. The hashing algorithm generates a hash value which is used to
+store every row of the left relation in the proper position in a `hash table`_.
+
+Then the rows of the right relation are read one-by-one and the same hashing
+algorithm is applied on the fields that participate in the join condition. The
+generated hash value is used to make a lookup in the `hash table`_. If no entry
+is found, the row is skipped and the processing continues with the next row
+from the right relation. If an entry is found, the join condition is validated
+(handling hash collisions) and on successful validation the combined tuple of
+left and right relation is returned.
+
+
+.. _joins_block_hash_join:
+
+Block Hash Join
+...............
+
+The Hash Join algorithm requires a `hash table`_ containing all the rows of the
+left relation to be stored in memory. Therefore, depending on the size of the
+relation (number of rows) and the size of each row, the size of this hash table
+might exceed the available memory of the node executing the hash join. To
+resolve this limitation the rows of the left relation are loaded into the hash
+table in blocks.
+
+On every iteration the maximum available size of the `hash table`_ is
+calculated, based on the number of rows and size of each row of the table but
+also taking into account the available memory for query execution on the node
+by the use of the `circuit breaker`_. Once this block-size is calculated the
+rows of the left relation are processed and inserted into the `hash table`_
+until the block-size is reached. The operation then starts reading the rows of
+the right relation, process them one-by-one and performs the lookup and the
+join condition validation. Once all rows from the right relation are processed
+the `hash table`_ is re-initialized based on a new calculation of the block
+size and a new iteration starts until all rows of the left relation are
+processsed.
+
+With this algorithm the memory limitation is raised in expense of having to
+iterate over the rows of the right table multiple times.
```

Comment:
I think somewhere we should highlight that the `block hash join` will always be used rather than the normal `hash join`. Also maybe mention that we will try use a `distributed block hash join` when possible (no limit/offset).